spring-kafka/src/main/java/org/springframework/kafka/retrytopic/DeadLetterPublishingRecovererFactory.java

```diff
@@ -185,11 +185,39 @@ private TopicPartition resolveDestination(ConsumerRecord<?, ?> cr, Exception e)
 				? "none"
 				: nextDestination.getDestinationName()));
 
+		logException(e, cr, nextDestination);
+
 		return nextDestination.isNoOpsTopic()
 					? null
 					: resolveTopicPartition(cr, nextDestination);
 	}
 
+	private void logException(Exception e, ConsumerRecord<?, ?> cr, DestinationTopic nextDestination) {
+		if (nextDestination.isDltTopic()) {
+			LOGGER.error(e, () -> getErrorMessage(cr)
+					+ "Sending to DLT with name " + nextDestination.getDestinationName() + ".");
+		}
+		else if (nextDestination.isNoOpsTopic()) {
+			LOGGER.error(e, () -> getErrorMessage(cr)
+					+ "No further action will be taken with this record.");
+		}
+		else {
+			LOGGER.debug(e, () -> "Error processing record: " + getRecordInfo(cr) + " at topic " + cr.topic() + ". "
+					+ "Sending to retry topic " + nextDestination.getDestinationName() + ".");
+		}
+	}
+
+	private String getErrorMessage(ConsumerRecord<?, ?> cr) {
+		return "Record: " + getRecordInfo(cr) + " threw an error at topic " + cr.topic() + " and won't be retried. ";
+	}
+
+	private String getRecordInfo(ConsumerRecord<?, ?> cr) {
```

Comment:
Looks like all of these methods can be `static`

Reply:
Sure. But if you don't bother me asking: why?

Reply:
Memory and performance: https://www.simplilearn.com/tutorials/java-tutorial/static-keyword-in-java.

Reply:
Nice, thanks! I'll change them.